casr/src/util.rs

```diff
@@ -293,3 +284,532 @@ pub fn get_output(command: &mut Command, timeout: u64, error_on_timeout: bool) -
+        }
+
+        let mut casr_cmd = Command::new(tool);
+        casr_cmd.args(&args);
+        casr_cmd.envs(envs);
+
+        debug!("{:?}", casr_cmd);
+
+        // Get output
+        let casr_output = casr_cmd
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cmd:?}"))?;
+
+        if !casr_output.status.success() {
+            let err = String::from_utf8_lossy(&casr_output.stderr);
+            if err.contains("Timeout") {
+                let mut timeout_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = timeout_name.find('-') {
+                    timeout_name.replace_range(..idx, "timeout");
+                }
+                let timeout_path = output_dir.join("timeout").join(timeout_name);
+                if fs::copy(&self.path, timeout_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Out of memory") {
+                let mut oom_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = oom_name.find('-') {
+                    oom_name.replace_range(..idx, "oom");
+                }
+                let oom_path = output_dir.join("oom").join(oom_name);
+                if fs::copy(&self.path, oom_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Program terminated (no crash)") {
+                warn!("{}: No crash on input {}", tool_name, self.path.display());
+            } else {
+                error!("{} for input: {}", err.trim(), self.path.display());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Perform analysis
+///
+/// # Arguments
+///
+/// `matches` - tool arguments
+///
+/// `crashes` - set of crashes, specified as a CrashInfo structure
+///
+/// `tool` - tool that generates reports
+///
+/// `gdb_argv` - arguments for casr-gdb
+pub fn generate_reports(
+    matches: &clap::ArgMatches,
+    crashes: &HashMap<String, CrashInfo>,
+    tool: &str,
+    gdb_argv: &Vec<String>,
+) -> Result<()> {
+    // Get tool paths
+    let casr_tool = get_path(tool)?;
+    let casr_gdb = get_path("casr-gdb")?;
+    let casr_cluster = get_path("casr-cluster")?;
+    // Get timeout
+    let timeout = if let Some(timeout) = matches.get_one::<u64>("timeout") {
+        *timeout
+    } else {
+        0
+    };
+
+    let (output_dir, _, _) = initialize_dirs(matches)?;
+
+    let envs = if tool.eq("casr-python") {
+        // Get Atheris asan_with_fuzzer library path.
```

Comment:
This can be moved to a separate function to improve the readability of the code.

Reply:
done